### ISSUE TITLE
Model not reproducible on different hardware, but same platform (Docker)

### ISSUE THREAD

**Issue by Zethson** [pytorch/pytorch#38219]
## 🐛 Bug

A simple MNIST model is not reproducible on CPU nor GPU, even with fixed seeds and deterministic CUDA settings enabled.
Basically, it is reproducible on a single hardware (e.g. my laptop), but does yield different results on another hardware (which by itself has different, but reproducible results).

e.g. Laptop: 
```
Run 1:
Test set: Average loss: 0.0467, Accuracy: 9917/10000 (99%)
GPU Run Time: 2313.047951698303 seconds
Run 2:
Test set: Average loss: 0.0467, Accuracy: 9917/10000 (99%)
GPU Run Time: 2231.947425365448 seconds
```

Cluster:
```
Run 1:
Test set: Average loss: 0.0538, Accuracy: 9914/10000 (99%)
GPU Run Time: 2147.833117246628 seconds
Run 2:
Test set: Average loss: 0.0538, Accuracy: 9914/10000 (99%)
GPU Run Time: 2385.5367307662964 seconds
```

## To Reproduce

1. Run simple MNIST from here: https://github.com/Zethson/nextflow_machine_learning/blob/master/bin/train_mnist_pytorch.py
1. Run it through this container: https://github.com/Zethson/nextflow_machine_learning/blob/master/Dockerfile_pytorch
1. With this env: https://github.com/Zethson/nextflow_machine_learning/blob/master/pytorch_environment.yml

## Expected behavior

I expect the model to be reproducible across hardware, when run in Docker and Singularity with all reproducibility settings switched on.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Nvidia Docker Cuda 10.02
 - How you installed PyTorch (`conda`, `pip`, source): Conda
 - Python version: 3.7
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 1050M and Tesla K80
 - Any other relevant information: Everything lives here https://github.com/Zethson/nextflow_machine_learning

## Additional context

Happy to answer any further questions.

**Comment by ngimel** [comment]
This is expected behavior, we don't guarantee bitwise accuracy for different hardware.

**Comment by Zethson** [comment]
@ngimel 
Thanks. Is there any information on that? Any docs or the origin of this behavior?

I could not find anything.

**Comment by ngimel** [comment]
https://pytorch.org/docs/stable/notes/randomness.html#reproducibility

**Comment by Zethson** [comment]
@ngimel I've read that several times already. There's not really a mention on hardware impact.

**Comment by ngimel** [comment]
```
Completely reproducible results are not guaranteed across PyTorch releases, individual commits or different platforms. Furthermore, results need not be reproducible between CPU and GPU executions, even when using identical seeds.
```
Different platforms means different hardware.

**Comment by Zethson** [comment]
I interpreted that as operating system, but all right.

Still, sorry to bother you. What are the reasons behind them? Trying to learn more about the sources and the design reasons behind them.

**Comment by ngimel** [comment]
Vendor libraries and compilers that pytorch relies on don't quarantee bitwise accurate results for different hardware they run on.

**Comment by Zethson** [comment]
Thanks. Final question (won't bother you about this anymore then - pinky promise):

Are there any plans on replacing the vendor libraries and moving to a different compiler for reproduciblity reasons in the future?

XGBoost for example is hardware reproducible and it also makes usage of several highly optimized libraries. 

There are a manifold of reasons, one obvious one being legal guarantees for model performance (e.g. computer vision tasks for face detection).
